internal/httpfile/httpfile.go

```diff
@@ -0,0 +1,429 @@
+	defer close(ra.done)
+	resp, err := client.Do(req)
+	if resp != nil {
+		ra.header = resp.Header
+		ra.statusCode = resp.StatusCode
+	}
+	defer func() {
+		if ra.err != nil && resp.Body != nil {
+			resp.Body.Close()
+		}
+	}()
+	if err != nil {
+		ra.err = err
+		return
+	}
+	switch resp.StatusCode {
+	case http.StatusNotFound:
+		ra.err = ErrNotFound
+	case http.StatusPreconditionFailed:
+		ra.err = errChanged
+	case http.StatusOK, http.StatusPartialContent:
+		length, err := f.checkContentRange(ra.p0, ra.p1, resp)
+		if err != nil {
+			ra.err = err
+			return
+		}
+		ra.length = length
+		ra.r = resp.Body
+	default:
+		ra.err = fmt.Errorf("unexpected response status %v", resp.StatusCode)
+	}
+}
+
+const noSkip = 64 * 1024
+
+// readAt tries to read data into buf at the given offset.
+// It differs from the ioutil.ReaderAt contract in that
+// it does not attempt to read the entire buffer when
+// less data is immediately available.
+func (ra *readerAhead) readAt(buf []byte, p0 int64) (int, error) {
+	if p0 < ra.p0 || p0 >= ra.p1 || p0 > ra.p0+noSkip {
+		return 0, errOutOfRange
+	}
+	<-ra.done
+	if ra.err != nil {
+		return 0, ra.err
+	}
+	if p0 > ra.p0 {
+		n, err := io.CopyN(ioutil.Discard, ra.r, p0-ra.p0)
+		ra.p0 += int64(n)
+		if err != nil {
+			if err == io.EOF {
+				return 0, io.ErrUnexpectedEOF
+			}
+			return 0, err
+		}
+	}
+	n, err := ra.r.Read(buf)
+	if err == io.EOF && n > 0 {
+		// Readers are entitled to return EOF at the end
+		// of a file, even when the read has been fulfilled.
+		// In our case, we only want to return EOF
+		// if there really are no more bytes to be had
+		// so suppress the EOF in that case.
+		err = nil
+	}
+	ra.p0 += int64(n)
+	return n, err
+}
+
+func (ra *readerAhead) remaining() int64 {
+	return ra.p1 - ra.p0
+}
+
+func (ra *readerAhead) close() {
+	go func() {
+		<-ra.done
+		if ra.r != nil {
+			ra.r.Close()
+		}
+	}()
+}
+
+func (f *File) newRequest(p0, p1 int64) *Request {
+	header := make(http.Header)
+	req := &Request{
+		Method: "GET",
+		Header: header,
+	}
+	switch {
+	case p0 == p1:
+		// No bytes requested - no need for a range request.
+		req.Method = "HEAD"
+	case p1 == infinity && p0 == 0:
+		// We want the whole thing - no need for a range request.
+	case p1 == infinity:
+		// Indefinite range not starting from the beginning.
+		// This case is here just for completeness - it
+		// doesn't happen in practice.
+		header.Set("Range", fmt.Sprintf("bytes=%d-", p0))
+	default:
+		header.Set("Range", fmt.Sprintf("bytes=%d-%d", p0, p1-1))
+	}
+	if f.etag != "" {
+		// Ensure that the read fails if the file has
+		// been written to since we opened it.
+		header.Set("If-Match", f.etag)
+	}
+	return req
+}
+
+// checkContentRange checks that the content range in the response is as
+// expected. If a server doesn't support range requests for whatever
+// reason, this should catch it.
+//
+// It returns the length of the entire
```

Comment:
I refactored this function into contentRangeFromResponse, which seems to
make the semantics clearer. The comment has gone.